<a href="https://colab.research.google.com/github/KdrDrn-C8301Sam/01-DAwPy-NumPy/blob/main/01_01_Preparing_The_Data_For_Analysis(Student_C8301_Sam).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

___

<p style="text-align: center;"><img src="https://docs.google.com/uc?id=1lY0Uj5R04yMY3-ZppPWxqCr5pvBLYPnV" class="img-fluid" alt="CLRSWY"></p>

___

<h1><p style='text-align: center;'>Traffic Police Stops<br>Part - 1</p><h1> - Preparing The Data For Analysis <img src='https://docs.google.com/uc?id=17CPCwi3_VvzcS87TOsh4_U8eExOhL6Ki' class='img-fluid' alt='CLRSWY' width='200' height='100'> 

Before beginning your analysis, it is critical that you first examine and clean the dataset, to make working with it a more efficient process. You will practice fixing data types, handling missing values, and dropping columns and rows while learning about the Stanford Open Policing Project dataset.

***

## Examining the dataset

You'll be analyzing a dataset of traffic stops in Rhode Island that was collected by the Stanford Open Policing Project.

Before beginning your analysis, it's important that you familiarize yourself with the dataset. You'll read the dataset into pandas, examine the first few rows, and then count the number of missing values.

**INSTRUCTIONS**

*   Import related libraries with their alias.
*   Read the file RI_cleaned.csv into a DataFrame named ``ri``
*   Examine the first 5 rows of the ``DataFrame`` (known as the ``"head"``).
*   Examine the last 5 rows of the ``DataFrame`` (known as the ``"tail"``).
*   Examine the ``DataFrame``'s ``"columns"``.
*   Count the number of missing values in each column: Use ``.isnull()`` to check which ``DataFrame`` elements are missing, and then take the ``.sum()`` to count the number of ``True`` values in each column.

In [37]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

In [38]:
pd.set_option('display.max_columns', None) # to see all columns

In [39]:
ri = pd.read_csv("RI_cleaned.csv")

C:\Users\hbpbn\AppData\Local\Temp\ipykernel_11016\3903740968.py:1: DtypeWarning: Columns (8,16) have mixed types. Specify dtype option on import or set low_memory=False.
  ri = pd.read_csv("RI_cleaned.csv")


In [40]:
ri.head()

id state   stop_date stop_time location_raw  county_name  \
0  RI-2005-00001    RI  2005-01-02     01:55      Zone K1          NaN   
1  RI-2005-00002    RI  2005-01-02     20:30      Zone X4          NaN   
2  RI-2005-00003    RI  2005-01-04     11:30      Zone X1          NaN   
3  RI-2005-00004    RI  2005-01-04     12:55      Zone X4          NaN   
4  RI-2005-00005    RI  2005-01-06     01:30      Zone X4          NaN   

   county_fips  fine_grained_location police_department driver_gender  \
0          NaN                    NaN               600             M   
1          NaN                    NaN               500             M   
2          NaN                    NaN                 0           NaN   
3          NaN                    NaN               500             M   
4          NaN                    NaN               500             M   

   driver_age_raw  driver_age driver_race_raw driver_race  \
0          1985.0        20.0               W       White   
1          1987.0        18.0               W       White   
2             NaN         NaN             NaN         NaN   
3          1986.0        19.0               W       White   
4          1978.0        27.0               B       Black   

                    violation_raw  violation search_conducted search_type_raw  \
0                        Speeding   Speeding            False             NaN   
1                        Speeding   Speeding            False             NaN   
2                             NaN        NaN            False             NaN   
3  Equipment/Inspection Violation  Equipment            False             NaN   
4  Equipment/Inspection Violation  Equipment            False             NaN   

  search_type  contraband_found stop_outcome is_arrested stop_duration  \
0         NaN             False     Citation       False      0-15 Min   
1         NaN             False     Citation       False     16-30 Min   
2         NaN             False          NaN         NaN           NaN   
3         NaN             False     Citation       False      0-15 Min   
4         NaN             False     Citation       False      0-15 Min   

  out_of_state  drugs_related_stop district  
0        False               False  Zone K1  
1        False               False  Zone X4  
2          NaN               False  Zone X1  
3        False               False  Zone X4  
4        False               False  Zone X4

In [41]:
ri.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 509681 entries, 0 to 509680
Data columns (total 26 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   id                     509681 non-null  object 
 1   state                  509681 non-null  object 
 2   stop_date              509671 non-null  object 
 3   stop_time              509671 non-null  object 
 4   location_raw           509681 non-null  object 
 5   county_name            0 non-null       float64
 6   county_fips            0 non-null       float64
 7   fine_grained_location  0 non-null       float64
 8   police_department      509671 non-null  object 
 9   driver_gender          480584 non-null  object 
 10  driver_age_raw         480632 non-null  float64
 11  driver_age             478986 non-null  float64
 12  driver_race_raw        480608 non-null  object 
 13  driver_race            480608 non-null  object 
 14  violation_raw          480608 non-nu

In [42]:
ri.describe().T

count         mean         std   min     25%  \
county_name                 0.0          NaN         NaN   NaN     NaN   
county_fips                 0.0          NaN         NaN   NaN     NaN   
fine_grained_location       0.0          NaN         NaN   NaN     NaN   
driver_age_raw         480632.0  1970.509997  108.187159   0.0  1967.0   
driver_age             478986.0    33.982027   12.702864  15.0    23.0   

                          50%     75%     max  
county_name               NaN     NaN     NaN  
county_fips               NaN     NaN     NaN  
fine_grained_location     NaN     NaN     NaN  
driver_age_raw         1980.0  1987.0  8801.0  
driver_age               31.0    43.0    99.0

In [43]:
ri.describe(include="O").T

count  unique                 top    freq
id                 509681  509681       RI-2005-00001       1
state              509681       1                  RI  509681
stop_date          509671    3809          2006-05-21     302
stop_time          509671    1440               10:00    1763
location_raw       509681       7             Zone X4  135349
police_department  509671      73                 500  113927
driver_gender      480584       2                   M  349446
driver_race_raw    480608       7                   W  344734
driver_race        480608       5               White  344734
violation_raw      480608      12            Speeding  268744
violation          480608       6            Speeding  268744
search_conducted   509671       2               False  491909
search_type_raw     17762      60  Incident to Arrest    6998
search_type         17762      25  Incident to Arrest    6998
stop_outcome       480608       6            Citation  428388
is_arrested        480608       2               False  464005
stop_duration      480608       5            0-15 Min  386665
out_of_state       479800       2               False  321276
district           509681       7             Zone X4  135349

**Since the dataset has 509681 rows, you can work with the first 50000 rows if you wish.**

In [44]:
ri = ri[:50000]

In [45]:
ri.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 26 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     50000 non-null  object 
 1   state                  50000 non-null  object 
 2   stop_date              50000 non-null  object 
 3   stop_time              50000 non-null  object 
 4   location_raw           50000 non-null  object 
 5   county_name            0 non-null      float64
 6   county_fips            0 non-null      float64
 7   fine_grained_location  0 non-null      float64
 8   police_department      50000 non-null  object 
 9   driver_gender          48010 non-null  object 
 10  driver_age_raw         48029 non-null  float64
 11  driver_age             47792 non-null  float64
 12  driver_race_raw        48012 non-null  object 
 13  driver_race            48012 non-null  object 
 14  violation_raw          48012 non-null  object 
 15  vi

In [46]:
ri.tail()

id state   stop_date stop_time location_raw  county_name  \
49995  RI-2006-35917    RI  2006-08-08     22:45      Zone K3          NaN   
49996  RI-2006-35918    RI  2006-08-08     22:45      Zone K3          NaN   
49997  RI-2006-35919    RI  2006-08-08     22:53      Zone X4          NaN   
49998  RI-2006-35920    RI  2006-08-08     23:00      Zone K1          NaN   
49999  RI-2006-35921    RI  2006-08-08     23:00      Zone K3          NaN   

       county_fips  fine_grained_location police_department driver_gender  \
49995          NaN                    NaN               300             M   
49996          NaN                    NaN               300             F   
49997          NaN                    NaN               500             M   
49998          NaN                    NaN               600             F   
49999          NaN                    NaN               300             M   

       driver_age_raw  driver_age driver_race_raw driver_race  \
49995          1973.0        33.0               B       Black   
49996          1971.0        35.0               B       Black   
49997          1952.0        54.0               W       White   
49998          1982.0        24.0               W       White   
49999          1965.0        41.0               W       White   

                        violation_raw  violation search_conducted  \
49995  Equipment/Inspection Violation  Equipment            False   
49996                        Speeding   Speeding            False   
49997                        Speeding   Speeding            False   
49998                        Speeding   Speeding            False   
49999                        Speeding   Speeding            False   

      search_type_raw search_type  contraband_found stop_outcome is_arrested  \
49995             NaN         NaN             False     Citation       False   
49996             NaN         NaN             False     Citation       False   
49997             NaN         NaN             False     Citation       False   
49998             NaN         NaN             False     Citation       False   
49999             NaN         NaN             False     Citation       False   

      stop_duration out_of_state  drugs_related_stop district  
49995      0-15 Min        False               False  Zone K3  
49996      0-15 Min         True               False  Zone K3  
49997     16-30 Min         True               False  Zone X4  
49998      0-15 Min        False               False  Zone K1  
49999     16-30 Min        False               False  Zone K3

In [47]:
ri.sample(5)

id state   stop_date stop_time location_raw  county_name  \
32695  RI-2006-18617    RI  2006-04-23     14:00      Zone K2          NaN   
6940   RI-2005-06941    RI  2005-11-12     12:45      Zone K1          NaN   
41101  RI-2006-27023    RI  2006-06-13     09:15      Zone X1          NaN   
46137  RI-2006-32059    RI  2006-07-16     11:40      Zone K3          NaN   
12630  RI-2005-12631    RI  2005-12-20     10:40      Zone K1          NaN   

       county_fips  fine_grained_location police_department driver_gender  \
32695          NaN                    NaN               900             M   
6940           NaN                    NaN               600             M   
41101          NaN                    NaN               000           NaN   
46137          NaN                    NaN               300             M   
12630          NaN                    NaN               600             F   

       driver_age_raw  driver_age driver_race_raw driver_race violation_raw  \
32695          1943.0        63.0               W       White      Speeding   
6940           1956.0        49.0               W       White      Speeding   
41101             NaN         NaN             NaN         NaN           NaN   
46137          1979.0        27.0               H    Hispanic      Speeding   
12630          1969.0        36.0               I       Asian      Speeding   

      violation search_conducted search_type_raw search_type  \
32695  Speeding            False             NaN         NaN   
6940   Speeding            False             NaN         NaN   
41101       NaN            False             NaN         NaN   
46137  Speeding            False             NaN         NaN   
12630  Speeding            False             NaN         NaN   

       contraband_found stop_outcome is_arrested stop_duration out_of_state  \
32695             False     Citation       False      0-15 Min        False   
6940              False     Citation       False      0-15 Min        False   
41101             False          NaN         NaN           NaN          NaN   
46137             False     Citation       False      0-15 Min        False   
12630             False     Citation       False      0-15 Min         True   

       drugs_related_stop district  
32695               False  Zone K2  
6940                False  Zone K1  
41101               False  Zone X1  
46137               False  Zone K3  
12630               False  Zone K1

In [48]:
ri.columns

Index(['id', 'state', 'stop_date', 'stop_time', 'location_raw', 'county_name',
       'county_fips', 'fine_grained_location', 'police_department',
       'driver_gender', 'driver_age_raw', 'driver_age', 'driver_race_raw',
       'driver_race', 'violation_raw', 'violation', 'search_conducted',
       'search_type_raw', 'search_type', 'contraband_found', 'stop_outcome',
       'is_arrested', 'stop_duration', 'out_of_state', 'drugs_related_stop',
       'district'],
      dtype='object')

In [49]:
ri.isnull().sum()

id                           0
state                        0
stop_date                    0
stop_time                    0
location_raw                 0
county_name              50000
county_fips              50000
fine_grained_location    50000
police_department            0
driver_gender             1990
driver_age_raw            1971
driver_age                2208
driver_race_raw           1988
driver_race               1988
violation_raw             1988
violation                 1988
search_conducted             0
search_type_raw          47988
search_type              47988
contraband_found             0
stop_outcome              1988
is_arrested               1988
stop_duration             1988
out_of_state              2203
drugs_related_stop           0
district                     0
dtype: int64

In [50]:
for col in ri.select_dtypes(include='O'):
    print(col)
    print('--'*10)
    print(ri[col].value_counts())
    print('**'*20)

id
--------------------
RI-2005-00001    1
RI-2006-19272    1
RI-2006-19250    1
RI-2006-19251    1
RI-2006-19252    1
                ..
RI-2006-02590    1
RI-2006-02591    1
RI-2006-02592    1
RI-2006-02593    1
RI-2006-35921    1
Name: id, Length: 50000, dtype: int64
****************************************
state
--------------------
RI    50000
Name: state, dtype: int64
****************************************
stop_date
--------------------
2006-05-21    302
2006-06-13    293
2005-11-05    290
2006-02-09    277
2006-03-22    276
             ... 
2005-06-06      1
2005-06-04      1
2005-05-21      1
2005-05-11      1
2005-06-18      1
Name: stop_date, Length: 378, dtype: int64
****************************************
stop_time
--------------------
10:00    433
09:00    420
23:30    388
11:00    374
08:30    370
        ... 
04:21      1
02:43      1
03:46      1
05:31      1
04:12      1
Name: stop_time, Length: 1376, dtype: int64
****************************************
location_r

In [51]:
for col in ri.select_dtypes(include='O'):
    print(col)
    print('--'*10)
    print(ri[col].unique())
    print('**'*20)

id
--------------------
['RI-2005-00001' 'RI-2005-00002' 'RI-2005-00003' ... 'RI-2006-35919'
 'RI-2006-35920' 'RI-2006-35921']
****************************************
state
--------------------
['RI']
****************************************
stop_date
--------------------
['2005-01-02' '2005-01-04' '2005-01-06' '2005-01-12' '2005-01-18'
 '2005-01-23' '2005-01-24' '2005-02-09' '2005-02-11' '2005-02-17'
 '2005-02-20' '2005-02-24' '2005-03-05' '2005-03-14' '2005-03-16'
 '2005-03-19' '2005-03-23' '2005-03-29' '2005-03-30' '2005-04-01'
 '2005-04-04' '2005-04-24' '2005-05-11' '2005-05-21' '2005-06-04'
 '2005-06-06' '2005-06-08' '2005-06-18' '2005-07-06' '2005-07-11'
 '2005-07-13' '2005-07-14' '2005-07-18' '2005-07-19' '2005-07-20'
 '2005-07-24' '2005-07-28' '2005-07-31' '2005-08-07' '2005-08-08'
 '2005-08-09' '2005-08-17' '2005-08-18' '2005-08-24' '2005-08-27'
 '2005-08-28' '2005-08-29' '2005-09-01' '2005-09-03' '2005-09-07'
 '2005-09-11' '2005-09-15' '2005-09-16' '2005-09-17' '2005-09-18'


***

## Dropping columns

Often, a DataFrame will contain columns that are not useful to your analysis. Such columns should be dropped from the ``DataFrame``, to make it easier for you to focus on the remaining columns.

You'll drop the ``county_name`` column because it only contains missing values, and you'll drop the ``state`` column because all of the traffic stops took place in one state (Rhode Island). Thus, these columns can be dropped because they contain no useful information.

**INSTRUCTIONS**

*   Examine the ``DataFrame``'s shape to find out the number of rows and columns.

*   Drop the columns that almost consist of missing values. 

*   Examine the ``.shape`` again to verify that there are now two fewer columns.

In [52]:
ri.shape

(50000, 26)

In [53]:
ri.drop(columns=['county_name', 'state', 'county_fips' , 'fine_grained_location', 'search_type_raw','search_type'], inplace=True)
ri.shape

(50000, 20)

***

## Dropping rows

When you know that a specific column will be critical to your analysis, and only a small fraction of rows are missing a value in that column, it often makes sense to remove those rows from the dataset.

During this course, the ``driver_gender`` column will be critical to many of your analyses. Because only a small fraction of rows are missing ``driver_gender``, we'll drop those rows from the dataset.

**INSTRUCTIONS**

*   Count the number of missing values in each column.

*   Drop all rows that are missing ``driver_gender`` by passing the column name to the subset parameter of ``.dropna()``.
*   Count the number of missing values in each column again, to verify that none of the remaining rows are missing ``driver_gender``.
*   Examine the ``DataFrame``'s ``.shape`` to see how many rows and columns remain.

In [54]:
ri.isnull().sum()

id                       0
stop_date                0
stop_time                0
location_raw             0
police_department        0
driver_gender         1990
driver_age_raw        1971
driver_age            2208
driver_race_raw       1988
driver_race           1988
violation_raw         1988
violation             1988
search_conducted         0
contraband_found         0
stop_outcome          1988
is_arrested           1988
stop_duration         1988
out_of_state          2203
drugs_related_stop       0
district                 0
dtype: int64

In [55]:
ri.dropna(subset=['driver_gender'], inplace=True)

In [56]:
ri.shape

(48010, 20)

In [57]:
ri.isnull().sum()

id                      0
stop_date               0
stop_time               0
location_raw            0
police_department       0
driver_gender           0
driver_age_raw          0
driver_age            232
driver_race_raw         0
driver_race             0
violation_raw           0
violation               0
search_conducted        0
contraband_found        0
stop_outcome            0
is_arrested             0
stop_duration           0
out_of_state          215
drugs_related_stop      0
district                0
dtype: int64

***

## Fixing a data type

We know that the ``is_arrested`` column currently has the ``object`` data type. In this exercise, we'll change the data type to ``bool``, which is the most suitable type for a column containing ``True`` and ``False`` values.

Fixing the data type will enable us to use mathematical operations on the ``is_arrested`` column that would not be possible otherwise.

**INSTRUCTIONS**

*   Examine the values of the ``is_arrested`` column to verify that it contains ``True`` and ``False`` values.
*   Check the current data type of ``is_arrested``.
*   Use the ``.astype()`` method to convert ``is_arrested`` to a ``bool`` column.
*   Check the new data type of ``is_arrested``, to confirm that it is now a ``bool`` column.

In [58]:
ri.is_arrested.value_counts()

False    45522
True      2488
Name: is_arrested, dtype: int64

In [59]:
ri.is_arrested.dtype

dtype('O')

In [66]:
ri = ri.astype({'is_arrested' : bool})

In [68]:
ri.is_arrested.dtype

dtype('bool')

***

## Combining object columns

Currently, the date and time of each traffic stop are stored in separate object columns: ``stop_date`` and ``stop_time``.

You'll combine these two columns into a single column, and then convert it to ``datetime`` format. This will enable convenient date-based attributes that we'll use later in the course.

**INSTRUCTIONS**

*    Get the first row of the data frame and see the dtypes of ``stop_date`` and ``stop_time`` columns.
*    Use a string method to concatenate ``stop_date`` and ``stop_time`` (separated by a space), and store the result in ``combined``.
*    Convert ``combined`` to ``datetime`` format, and store the result in a new column named ``stop_datetime``.
*    Drop ``stop_date`` and ``stop_time`` columns from dataframe and check after dropping by getting first row.
*    Examine the ``DataFrame`` ``.dtypes`` to confirm that ``stop_datetime`` is a datetime column.

***

The last step that you'll take in this chapter is to set the ``stop_datetime`` column as the ``DataFrame``'s index. By replacing the default index with a ``DatetimeIndex``, you'll make it easier to analyze the dataset by date and time, which will come in handy later in the course.

**INSTRUCTIONS**

*   First check out the shape of the dataframe.
*   Set ``stop_datetime`` as the ``DataFrame`` index.
*   Examine the index to verify that it is a ``DatetimeIndex``.
*   Check again the shape of the dataframe to confirm that ``stop_datetime`` is no longer one of the columns. 
*   Examine the ``DataFrame`` columns to confirm that ``stop_datetime`` is no longer one of the columns.

___

<p style="text-align: center;"><img src="https://docs.google.com/uc?id=1lY0Uj5R04yMY3-ZppPWxqCr5pvBLYPnV" class="img-fluid" alt="CLRSWY"></p>

___

**- Please save the edits made so far to a new csv file with the name RI_Part_1 to use it in the next step.**

**- Load and read new csv file.**

**- Check the first five rows.***

**- Check the shape of the dataframe.**